spec/ics-3-connection-semantics/README.md

```diff
@@ -0,0 +1,245 @@
+- `P` is the unbonding period of chain `P`, in units of time
+- `dt(a, b)` is the time difference between events `a` and `b` 
+
+Note that of all these, only `H_h` defines a signature and is thus attributable.
+
+#### Requirements
+
+(reference ICS 2: consensus primitives)
+
+To facilitate an IBC connection, the two blockchains must provide the following proofs:
+
+1. Given a trusted `H_h` and `C_h` and an attributable update message `U_h`,  
+   it is possible to prove `H_h'` where `C_h' == C_h` and `dt(now, H_h) < P`
+2. Given a trusted `H_h` and `C_h` and an attributable change message `X_h`,  
+   it is possible to prove `H_h'` where `C_h' /= C_h` and `dt(now, H_h) < P`
+3. Given a trusted `H_h` and a Merkle proof `M_kvh` it is possible to prove `V_kh`
+
+It is possible to make use of the structure of BFT consensus to construct extremely lightweight and provable messages `U_h'` and `X_h'`. The implementation of these requirements with Tendermint consensus is defined in [Appendix E](appendices.md#appendix-e-tendermint-header-proofs). Another algorithm able to provide equally strong guarantees (such as Casper) is also compatible with IBC but must define its own set of update and change messages.
+
+The Merkle proof `M_kvh` is a well-defined concept in the blockchain space, and provides a compact proof that the key value pair `(k, v)` is consistent with a Merkle root stored in `H_h`. Handling the case where `k` is not in the store requires a separate proof of non-existence, which is not supported by all Merkle stores. Thus, we define the proof only as a proof of existence. There is no valid proof for missing keys, and we design the algorithm to work without it. 
+
+Blockchains supporting IBC must implement Merkle proof verification:
+
+`valid(H_h, M_kvh) ⇒ true | false`
+
+#### Lifecycle
+
+##### Opening a connection
+
+All proofs require an initial `H_h` and `C_h` for some `h`, where `dt(now, H_h) < P`.
+
+Establishing a bidirectional initial root-of-trust between the two blockchains (`A` to `B` and `B` to `A`) — `H_ah` and `C_ah` stored on chain `B`, and `H_bh` and `C_bh` stored on chain `A` — is necessary before any IBC packets can be sent.
+
+Any header may be from a malicious chain (e.g. shadowing a real chain state with a fake validator set), so a subjective decision is required before establishing a connection. This can be performed permissionlessly, in which case users later utilizing the IBC channel must check the root-of-trust themselves, or authorized by on-chain governance for additional assurance.
+
+*THREE WAY HANDSHAKE OUTLINE*
+
+###### OPENTRY
+
+First datagram, A -> B
+
+Provided input:
+- Commitment to root-of-trust for B
+
+On A:
+- Reserve connection slot
+- Prove reserved connection slot (is this necessary?)
+- Prove commitment to root-of-trust for B
+
+Data:
+- Root-of-trust for A
+- Connection slot identifier on A
+- Desired connection slot identifier on B
```

Comment:
Why is this necessary? And if its part of the relay data shouldnt it be committed on ChainA already ?

Reply:
The two chains need not use the same identifier (this is *not* necessarily an identifier like "Cosmos Hub", it's just a key for the connection, it could be an unsigned integer for example). The handshake protocol allows the two chains to verify the identifier the other is using, in case they do care about what it is or want a common reference.